046-analyse-broker-metrics-before-rolling-kafka.md

```diff
@@ -0,0 +1,93 @@
+
+The following implementation details will describe behaviour of the new API in KafkaAgent and how KafkaRoller would interact with it and handle a broker that is in log recovery.
+
+### Implementation details
+
+**KafkaAgent**:
+
+- Add a web server to accept SSL connections on port 8443.
+- Use Jetty for the web server since Kafka already uses it therefore avoids an additional dependency.
+- Reuse the broker's own certificate that would be in the container for TLS authentication. 
+- Configure the server to require TLS client authentication. 
+- Expose an endpoint `v1/recovery-state` that includes the API version in the URI path.
+- Extend KafkaAgent to collect `kafka.log:type=LogManager,name=remainingLogsToRecover` and `kafka.log:type=LogManager,name=remainingSegmentsToRecover` metrics.
+- Use Jetty's native ServerHandler API to handle a `GET` request and return a JSON response that looks like this:
+```
+{
+  "brokerState": 2,
+  "remainingLogsToRecover": 123,
+  "remainingSegmentsToRecover": 456
+}
+```
+- Return HTTP 404 (Not Found) if a request is sent to an unknown version of the API e.g. `GET /v5/recovery-state` and `v5` does not exist.
+
+**KafkaRoller**:
+
+The proposed flow of KafkaRoller
+![The proposed flow of KafkaRoller](images/046-kafka-roller-new-flow.png)
+
+If a kafka pod does not become ready within the operational timeout (5 mins by default) take the following actions:
+1. Check if the maximum number of retries has been reached. If it has, fail the reconciliation.
+2. Ensure that the pod is not in a stuck state. The existing `isPodStuck()` can be used for this which returns true if the pod is in `CrashLoopBackOff`, `ImagePullBackOff`, `ContainerCreating` or `Pending` state. If the pod is stuck, continue with the current behaviour to retry rolling of the pod.
+3. Use the Pod DNS name to query the KafkaAgent API endpoint.
+4. If the API returns 503 (Service Unavailable), continue with the current behaviour to retry rolling of the pod.
+5. Use the cluster operator's client certificate and keys to authenticate with the KafkaAgent API server.
+6. Send `GET v1/recovery-state` request to the endpoint and check the value of `brokerState` from the response.
+7. If the `brokerState` value is `2` (`RECOVERY`), follow the new logic for handling log recovery.
+ 
+*Handling log recovery*
+
+8. Set `inLogRecovery` to true. (`inLogRecovery` is a new boolean that would be added to the `RestartContext`).
+9. Log the values of `remainingLogsToRecover` and `remainingSegmentsToRecover` from the request response in an `INFO` message.
+10. Continue with the current behaviour to retry rolling of the pod.
+11. On the next retry, check `inLogRecovery` inside the `restartIfNecessary()` function.
+12. When `inLogRecovery` is set to true, call `awaitReadiness()` which periodically checks and waits for readiness until the operational timeout reaches.
```

Comment:
We already waited for 5 minutes for the readiness before this. Do we want to wait again? We ideally want to make sure to not do too many timing out _waiting for readiness_ calls in one reconciliation. Looping the reconciliation is important because it helps to inform the user.

Reply:
We are not adding any additional wait here. On the timeout, it will simply set `inLogRecovery` to true and move on to the next retry. On the retry, it will wait for readiness right away because inLogRecovery is set to true. It already does that if `forceRestart` is set false. The retry logic on the timeout is not new.